src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -793,42 +816,50 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
             StringBuilder? sb = null;
             do
             {
-                int i = _charPos;
-                do
+                // Look for '\r' or \'n'.
+                ReadOnlySpan<char> charBufferSpan = _charBuffer.AsSpan(_charPos, _charLen - _charPos);
+                Debug.Assert(!charBufferSpan.IsEmpty, "ReadBuffer returned > 0 but didn't bump _charLen?");
+
+                int idxOfNewline = charBufferSpan.IndexOfAny('\r', '\n');
+                if (idxOfNewline >= 0)
                 {
-                    char ch = _charBuffer[i];
-                    // Note the following common line feed chars:
-                    // \n - UNIX   \r\n - DOS   \r - Mac
-                    if (ch == '\r' || ch == '\n')
+                    string retVal;
+                    if (sb is null)
                     {
-                        string s;
-                        if (sb != null)
-                        {
-                            sb.Append(_charBuffer, _charPos, i - _charPos);
-                            s = sb.ToString();
-                        }
-                        else
-                        {
-                            s = new string(_charBuffer, _charPos, i - _charPos);
-                        }
-                        _charPos = i + 1;
-                        if (ch == '\r' && (_charPos < _charLen || ReadBuffer() > 0))
+                        retVal = new string(charBufferSpan.Slice(0, idxOfNewline));
+                    }
+                    else
+                    {
+                        sb.Append(charBufferSpan.Slice(0, idxOfNewline));
+                        retVal = StringBuilderCache.GetStringAndRelease(sb);
+                    }
+
+                    char matchedChar = charBufferSpan[idxOfNewline];
+                    _charPos += idxOfNewline + 1;
+
+                    // If we found '\r', consume any immediately following '\n'.
+                    if (matchedChar == '\r')
+                    {
+                        if (_charPos < _charLen || ReadBuffer() > 0)
                         {
                             if (_charBuffer[_charPos] == '\n')
                             {
                                 _charPos++;
                             }
                         }
-                        return s;
                     }
-                    i++;
-                } while (i < _charLen);
 
-                i = _charLen - _charPos;
-                sb ??= new StringBuilder(i + 80);
-                sb.Append(_charBuffer, _charPos, i);
+                    return retVal;
+                }
+
+                // We didn't find '\r' or '\n'. Add it to the StringBuilder
+                // and loop until we reach a newline or EOF.
+
+                sb ??= StringBuilderCache.Acquire(charBufferSpan.Length + 80);
```

Comment:
Should `ValueStringBuilder` be used instead?
I'm asking because there's https://github.com/dotnet/runtime/pull/69683

Reply:
Yeah, I've been slowly trying to rid us of StringBuilderCache, as it doesn't share buffers with ArrayPool and ends up failing to cache once you hit the limit of ~350 chars.  The one advantage StringBuilder has in these situations is that it's a linked list of segments which won't end up on the LOH, but if the expected case here is 80-100 chars per line, it should be better to just use VSB, which can also use stack space and avoid hitting the ArrayPool at all for the vast majority case.  We'll only be able to use it for ReadLine, though, not ReadLineAsync.

Reply:
> Basically, I don't foresee better perf in the typical use case, and I do foresee worse perf in the extreme use case, when compared to a standard StringBuilder.

I would expect VSB to have at least slightly better perf in the typical use case, as it's able to use stack memory instead of paying for a TLS access.

But more importantly, for lines longer than 360 characters (which isn't that unusual in situations where each paragraph is a line), VSB will use the ArrayPool whereas StringBuilderCache will allocate new StringBuilders/underlying char[]s each time.

It's only in the extreme worst case scenario that we could end up creating LOH arrays with VSB where we wouldn't with StringBuilder, but even there, they'll be pooled with VSB whereas they won't be with StringBuilder.

Reply:
If we wanted to make a change to use VSB, I'd feel more comfortable if somebody submits that as a future change. It has potential to impact the performance profile of this method in edge cases in ways I don't feel I can correctly predict. We could also tackle the issue you said: figure out how to use pooling in both the sync and async cases at the same time.

Reply:
> If we wanted to make a change to use VSB, I'd feel more comfortable if somebody submits that as a future change

Ok.

> It has potential to impact the performance profile of this method in edge cases in ways I don't feel I can correctly predict.

Can you share an example edge case you're worried about?

Reply:
Extreme examples where we're talking buffers dozens or hundreds of MB in size. Will VSB fail in scenarios where StringBuilder would have succeeded, perhaps because StringBuilder's mechanism of using non-contiguous buffers works better in low-memory situations? Has VSB been battle-tested using large buffers? (I'm looking through its code and it looks like the `Grow` method has a functional bug when very large buffers come into play.)

Reply:
My comment on the `Grow` method is that I think the rental logic should look like this instead of how it looks currently.

```cs
int lengthToRent = (int)Math.Max(
    (uint)(_pos + additionalCapacityBeyondPos),
    Math.Min((uint)_chars.Length * 2, (uint)Array.MaxLength));

// Make sure to let Rent throw an exception if the caller has a bug and the desired capacity is negative
char[] poolArray = ArrayPool<char>.Shared.Rent(lengthToRent);
```

But it needs proper testing, and that's a discussion for a separate issue anyway. :)

Reply:
https://github.com/dotnet/runtime/pull/70942